src/cmdline/mod.rs

```diff
@@ -336,172 +250,669 @@ impl Cmdline {
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use linux_loader::cmdline::*;
+    /// let cl = Cmdline::new(100);
+    /// ```
+    /// [`Cmdline`]: struct.Cmdline.html
+    pub fn new(cmdline_capacity: usize) -> Result<Cmdline> {
+        if cmdline_capacity == 0 {
+            return Err(Error::InvalidCapacity);
+        }
+
+        Ok(Cmdline {
+            boot_args: BootArgs(String::new()),
+            init_args: InitArgs(String::new()),
+            cmdline_capacity,
+        })
+    }
+
+    /// Returns the BootArgs representation of the command line.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use linux_loader::cmdline::*;
+    /// let mut cl = Cmdline::new(100).unwrap();
+    /// assert!(cl.boot_args_mut().insert_key_value("a", "b").is_ok());
+    /// assert_eq!(cl.as_str().unwrap(), "a=b");
+    /// ```
+    pub fn boot_args_mut(&mut self) -> &mut BootArgs {
+        &mut self.boot_args
+    }
+
+    /// Returns the InitArgs representation of the command line.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use linux_loader::cmdline::*;
+    /// let mut cl = Cmdline::new(100).unwrap();
+    /// assert!(cl.boot_args_mut().insert_key_value("a", "b").is_ok());
+    /// assert!(cl.init_args_mut().insert_key_value("c", "d").is_ok());
+    /// assert_eq!(cl.as_str().unwrap(), "a=b -- c=d");
+    /// ```
+    pub fn init_args_mut(&mut self) -> &mut InitArgs {
+        &mut self.init_args
+    }
+
+    /// Returns the string representation of the command line without the nul terminator.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use linux_loader::cmdline::*;
+    /// let mut cl = Cmdline::new(100).unwrap();
+    /// assert!(cl.boot_args_mut().insert_str("foo").is_ok());
+    /// assert!(cl.init_args_mut().insert_str("bar").is_ok());
+    /// assert_eq!(cl.as_str().unwrap(), "foo -- bar");
+    /// ```
+    pub fn as_str(&self) -> Result<String> {
+        if self.boot_args.0.is_empty() {
+            return Err(Error::NoBootArgsInserted);
+        }
+
+        if self.init_args.0.is_empty() {
+            if self.boot_args.0.len() + 1 > self.cmdline_capacity {
+                // need space for the null terminator
+                Err(Error::TooLarge)
+            } else {
+                Ok(self.boot_args.0.to_string())
+            }
+        } else if self.boot_args.0.len() + self.init_args.0.len() + 5 > self.cmdline_capacity {
+            // need space for the null terminator and ' -- ' delimiter
+            Err(Error::TooLarge)
+        } else {
+            Ok(format!("{} -- {}", self.boot_args.0, self.init_args.0))
+        }
+    }
+}
+
+fn check_outside_double_quotes(slug: &str) -> bool {
+    slug.matches('\"').count() % 2 == 0
+}
+
+impl TryFrom<(usize, String)> for Cmdline {
+    type Error = String;
+
+    fn try_from(
+        (cmdline_capacity, cmdline_raw): (usize, String),
```

Comment:
If we want to have a capacity as well then I think implementing `TryFrom` is no longer a good idea because we have to use a tuple, so it's no longer clear how to initialize the `Cmdline`. In case we want to keep the capacity, we should just have a function on `Cmdline` so that we can have as named parameters a `value` and a `capacity`. I would recommend thought to drop `capacity` altogether and just check (as you're already doing) that the capacity of the generated string is less than the maximum allowed one by the linux kernel.

Reply:
Ok, implementing this way. I think that the cmdline_capacity param came from the fact that each arch has different cmdline_max sizes; we can allow the maximum of these values (4096)